### PR TITLE
Enhancement/2457 excessive analytics requests

### PR DESCRIPTION
**Pull request by asvinb**
## Summary

Addresses issue #2457 

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/googlesitekit/api/cache.js [hunk 1]
```diff
@@ -202,19 +203,28 @@ export const getItem = async ( key, cacheTimeToLive = null ) => {
...
+ * @param {*}       value          Value to store in the cache.
+ * @param {Object}  args           Optional object containing ttl, timestamp and isError keys.
+ * @param {number}  args.ttl       Validity of the cached item in seconds.
+ * @param {number}  args.timestamp Timestamp when the cached item was created.
+ * @param {boolean} args.isError   Whether the cached item is an error.
  * @return {Promise} A promise: resolves to `true` if the value was saved; `false` if not (usually because no storage method was available).
  */
-export const setItem = async ( key, value, _timestamp = undefined ) => {
+export const setItem = async ( key, value, {
+	ttl = 60 * 60,
```
Comment (eugene-manuilov) [inline assets/js/googlesitekit/api/cache.js:215]:
```suggestion
	ttl = 3600, // One hour.
```

File: assets/js/googlesitekit/api/cache.js [hunk 2]
```diff
@@ -202,19 +203,28 @@ export const getItem = async ( key, cacheTimeToLive = null ) => {
...
  *
- * @param {string} key        Name of cache key.
- * @param {*}      value      Value to store in the cache.
- * @param {number} _timestamp Timestamp to set as the cache data save time.
+ * @param {string}  key            Name of cache key.
+ * @param {*}       value          Value to store in the cache.
+ * @param {Object}  args           Optional object containing ttl, timestamp and isError keys.
+ * @param {number}  args.ttl       Validity of the cached item in seconds.
+ * @param {number}  args.timestamp Timestamp when the cached item was created.
+ * @param {boolean} args.isError   Whether the cached item is an error.
```
Comment (eugene-manuilov) [inline assets/js/googlesitekit/api/cache.js:211]:
```suggestion
 * @param {number}  [args.ttl]       Optional. Validity of the cached item in seconds.
 * @param {number}  [args.timestamp] Optional. Timestamp when the cached item was created.
 * @param {boolean} [args.isError]   Optional. Whether the cached item is an error.
```

File: assets/js/googlesitekit/api/cache.js [hunk 3]
```diff
@@ -161,29 +161,30 @@ export const getStorage = async () => {
...
+			const { timestamp, ttl, value, isError } = parsedData;
 
 			// Ensure a timestamp is found, otherwise this isn't a valid cache hit.
 			// (We don't check for a truthy `value`, because it could be legitimately
 			// false-y if `0`, `null`, etc.)
-			if ( parsedData.timestamp && (
-				cacheTimeToLive === null || // Ensure the cached data isn't too old.
-				Math.round( Date.now() / 1000 ) - parsedData.timestamp < cacheTimeToLive
+			if ( timestamp && (
+				ttl === null || ttl === undefined || // Ensure the cached data isn't too old.
```
Comment (eugene-manuilov) [inline assets/js/googlesitekit/api/cache.js:181]:
```suggestion
				! ttl || // Ensure the cached data isn't too old.
```

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Nice work, @asvinb! Just found a few minor things to improve and it should be good to go. Could you please check it?

File: includes/Core/Modules/Module.php [hunk 4]
```diff
@@ -831,15 +831,21 @@ protected function exception_to_error( Exception $e, $datapoint ) {
...
 		$reconnect_url = '';
 
 		if ( $e instanceof Google_Service_Exception ) {
+
 			$errors = $e->getErrors();
 			if ( isset( $errors[0]['message'] ) ) {
 				$message = $errors[0]['message'];
+
+				if ( preg_match( '#^Restricted metric\(s\):(.*)$#im', $message ) ) {
+					$cache_ttl = ( 10 * 60 );
```
[the remark this answers is not in the document]
Comment (asvinb) [reply includes/Core/Modules/Module.php]: Updated @eugene-manuilov

File: includes/Core/Modules/Module.php [hunk 5]
```diff
@@ -831,15 +831,21 @@ protected function exception_to_error( Exception $e, $datapoint ) {
...
 		$reason        = '';
 		$reconnect_url = '';
 
 		if ( $e instanceof Google_Service_Exception ) {
+
 			$errors = $e->getErrors();
 			if ( isset( $errors[0]['message'] ) ) {
 				$message = $errors[0]['message'];
+
+				if ( preg_match( '#^Restricted metric\(s\):(.*)$#im', $message ) ) {
```
[the remark this answers is not in the document]
Comment (asvinb) [reply includes/Core/Modules/Module.php]: Nice. Thanks!

File: assets/js/googlesitekit/api/cache.js (hunk 1 above)
Comment (asvinb) [reply assets/js/googlesitekit/api/cache.js]: 👍

File: assets/js/googlesitekit/api/cache.js (hunk 2 above)
Comment (asvinb) [reply assets/js/googlesitekit/api/cache.js]: 👍

File: assets/js/googlesitekit/api/cache.js (hunk 3 above)
Comment (asvinb) [reply assets/js/googlesitekit/api/cache.js]: 👍

**Review by eugene-manuilov — APPROVED** [review]
LGTM!

File: includes/Core/Modules/Module.php [hunk 6]
```diff
@@ -861,6 +867,10 @@ protected function exception_to_error( Exception $e, $datapoint ) {
 			'reason' => $reason,
 		);
 
+		if ( $cache_ttl ) {
+			$data['cacheTTL'] = $cache_ttl;
+		}
+
```
Comment (felixarntz) [inline includes/Core/Modules/Module.php:873]: While this works, the entire PHP logic needed here is specific to the Analytics API, so I don't think it is suitable to be added to the general `Module` class (also see "the `Analytics` exception handling function" reference in the ACs). Let's instead put these changes into the `Analytics` class:
* Modify the overridden `exception_to_error` method, which already includes the necessary check.
* Instead of immediately returning the `WP_Error` returned by the parent method, modify it to add the `cacheTTL` to the `$data` property of the `WP_Error` object based on the checks made.
    * Basically, the existing clause can be moved below the parent method call so that we can then within it update the `WP_Error` as needed before returning it.

**Review by felixarntz — CHANGES_REQUESTED** [review]
Mostly LGTM, however the PHP logic should not be in the general `Module` class since it is specific to Analytics (see ACs and IB).

File: includes/Core/Modules/Module.php (hunk 6 above)
Comment (asvinb) [reply includes/Core/Modules/Module.php]: Ah! Makes sense @felixarntz I thought we were going to do that for all "Restricted metrics..." error. The PR has been updated. Can you take a look @eugene-manuilov ?

**Review by eugene-manuilov — DISMISSED** [review]
Looks good!

File: includes/Modules/Analytics.php [hunk 7]
```diff
@@ -1299,15 +1299,33 @@ protected function has_access_to_property( $property_id ) {
...
 
-		return parent::exception_to_error( $e, $datapoint );
+		$error = parent::exception_to_error( $e, $datapoint );
+
+		if ( $cache_ttl && is_wp_error( $error ) ) {
+			$error->add_data(
+				array(
+					'cacheTTL' => $cache_ttl,
+				)
+			);
```
Comment (felixarntz) [inline includes/Modules/Analytics.php:1325]: This unconditionally overrides any existing `data` that is already present in the `$error` (e.g. the HTTP response code in `status` or the `errorReason`). We'll need to change it to manually alter the `$error->error_data` property if it already has content (modifying properties directly isn't great, but that's all that WordPress core allows here).

**Review by felixarntz — CHANGES_REQUESTED** [review]
Almost good to go, just one final comment to address.

**Review by felixarntz — APPROVED** [review]
Fixed the last bit, so this is good to go now!
